src/test_coroutines.cc

```diff
@@ -0,0 +1,136 @@
+// Copyright 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <coroutine>
+#include <queue>
+
+#include "coroutine.h"
+#include "doctest.h"
+
+namespace triton::perfanalyzer {
+
+// A simple coroutine that returns an integer.
+// It awaits for an Awaiter object and then returns 42,
+// meaning that it will resume twice before completing
+// as the coroutine is created in a suspended state.
+Coroutine<int>
+CoroutineTest()
+{
+  co_await Coroutine<int>::Awaiter{};
+  co_return 42;
+}
+
+TEST_CASE("coroutine:testing the Coroutine class")
+{
+  auto coroutine = CoroutineTest();
+
+  unsigned rounds = 0;
+  while (!coroutine.Done()) {
+    coroutine.Resume();
+    rounds++;
+  }
+
+  auto result = coroutine.Value();
+
+  CHECK(rounds == 2);
+  CHECK(result == 42);
+  CHECK(coroutine.Done());
+}
+
+// A simple coroutine that returns void.
+// This is the same as the previous test, but with a void return type,
+// in order to test the void specialization of the Coroutine class.
+Coroutine<>
+CoroutineVoidTest()
+{
+  co_await Coroutine<>::Awaiter{};
+}
+
+TEST_CASE("coroutine:testing the Coroutine class with void")
+{
+  auto coroutine = CoroutineVoidTest();
+
+  unsigned rounds = 0;
+  while (!coroutine.Done()) {
+    coroutine.Resume();
+    rounds++;
+  }
```

Comment:
Took me a few minutes of reading code to understand why rounds was 2. I think a test like this would have been quicker to gleam the understanding of while still validating the same behavior.

```suggestion
REQUIRE(!coroutine.Done());

coroutine.Resume(); // because coroutine is initially suspended
coroutine.Resume(); // because coroutine suspends at end

REQUIRE(coroutine.Done());
```

Reply:
I don't mind changing this, though it takes away a bit of the original goal of showcasing how a proto-scheduler works. Basically, the point of a scheduler is to run a coroutine until it's done, but there's a more complex example below anyway, so, sure.

Reply:
Sure feel free to keep it as a loop, just offering ideas where there's agreement